server/src/main/java/org/elasticsearch/health/node/DiskHealthIndicatorService.java

```diff
@@ -489,4 +149,264 @@ private HealthIndicatorDetails getDetails(boolean explain, Map<String, DiskHealt
             return builder.endObject();
         });
     }
+
+    /**
+     * The disk health analyzer takes into consideration the blocked indices and the health status of the all the nodes and calculates
+     * the different aspects of the disk indicator such as the overall status, the symptom, the impacts and the diagnoses.
+     */
+    static class DiskHealthAnalyzer {
+
+        private final ClusterState clusterState;
+        private final Set<String> blockedIndices;
+        private final Set<DiscoveryNode> dataNodes = new HashSet<>();
+        // In this context a master node, is a master node that cannot contain data.
+        private final Set<DiscoveryNode> masterNodes = new HashSet<>();
+        // In this context "other" nodes are nodes that cannot contain data and are not masters.
+        private final Set<DiscoveryNode> otherNodes = new HashSet<>();
+        private final Set<DiscoveryNodeRole> affectedRoles = new HashSet<>();
+        private final Set<String> indicesAtRisk;
+        private final HealthStatus healthStatus;
+
+        DiskHealthAnalyzer(Map<String, DiskHealthInfo> diskHealthByNode, Set<String> blockedIndices, ClusterState clusterState) {
+            this.clusterState = clusterState;
+            this.blockedIndices = blockedIndices;
+            HealthStatus mostSevereNodeStatus = HealthStatus.GREEN;
+            for (String nodeId : diskHealthByNode.keySet()) {
+                DiscoveryNode node = clusterState.getNodes().get(nodeId);
+                HealthStatus healthStatus = diskHealthByNode.get(nodeId).healthStatus();
+                // TODO #90213 update this only after we check that this health status indicates a problem.
+                if (mostSevereNodeStatus.value() < healthStatus.value()) {
+                    mostSevereNodeStatus = healthStatus;
+                }
+                if (node == null || healthStatus.indicatesHealthProblem() == false) {
+                    continue;
+                }
+                affectedRoles.addAll(node.getRoles());
+                if (node.canContainData()) {
+                    dataNodes.add(node);
+                } else if (node.isMasterNode()) {
+                    masterNodes.add(node);
```

Comment:
Do we not want to differentiate a red vs a yellow master node? 
Same for the other roles (ie. at least the diagnosis is different - "Disk is full" vs "Elasticsearch is running low on disk space" )

Reply:
Yes we can do that. I wasn't sure if it is worth the granularity but if it is making things more clear it's worth it. I will adjust it.

Reply:
So after some thought, I agree to do this for the non data nodes, especially because increasing their capacity actually requires to increase the capacity of the affected node. 

However, I think we should keep the data nodes grouped together because they work complementary to each other, adding a data node for example or increasing the disk space of a different node usually helps. What do you think about this?

Reply:
Sounds good, thanks Mary

Reply:
This can be done as a follow-up PR